doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,456 @@
+
+#### Read Then Write Statements
+
+`UPDATE`, `DELETE`, and `INSERT INTO SELECT` queries are all classified as ReadThenWrite statements, because they all
+first perform a read and then perform a write. Materialize does not currently support general purpose Read/Write
+transactions, but has a specific implementation for these types of statements.
+
+The first thing a ReadThenWrite statement does is acquire a lock on the Coordinator. This lock prevents all write-only
+transactions and other ReadThenWrite statements from executing until the current ReadThenWrite statement has fully
+completed. The lock does block any other operation. Then a timestamp is selected and the read is performed at that
+timestamp. Finally, the write behaves identically to a write-only transaction containing a single statement (a new
+timestamp is selected for the write).
+
+For more information you can read the design document for these
+statements: [https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20210715_table_update.md](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20210715_table_update.md)
+
+NOTE: There is active discussion on making these not require a global lock on the Coordinator to help with performance.
+
+### Timestamp Selection
+
+All reads and writes happen at a specific timestamp that is selected by Materialize. This section describes how
+Materialize selects timestamps. All timestamp selection happens on a single Coordinator thread, so no two timestamps can
+be selected concurrently. The [Strict Serializable](#Strict Serializable) section will propose some changes to achieve a
+consistency level of Strict Serializable.
+
+#### Since and Upper
+
+Every object maintains a range of timestamps, [`since`, `upper`). All reads to that object must take place at a
+timestamp greater than or equal to `since`. If a timestamp larger than `upper` is selected to read an object, the read
+will block until `upper` has advanced past the selected timestamp. All writes to that object must take place at a
+timestamp greater than or equal to `upper`. Throughout the lifetime of the object, both `since` and `upper` can
+increase, but they never decrease. This ensures that a read at timestamp `ts` will always return the same value, because
+if `ts` is readable, then all future writes must take place at a timestamp greater than `ts`.
+
+NOTE: When an object is first created the `upper` may be initialized below the `since`. This will quickly correct itself
+and, as mentioned above, all reads will block until the upper has advanced.
+
+#### User Tables
+
+The Coordinator maintains a global timestamp for all user tables. Every one second this timestamp is increased by one
+second and all tables are advanced to the new timestamp (reads and writes can also increase the timestamp and/or advance
+tables as described below). Advancing a table to `ts` causes its `upper` to be set to `ts`, and its `since` to be set
```

Comment:
`since` is advanced if there is not an open read-only transaction preventing compaction (compaction occurs when `since` advances)

Reply:
Added a couple of sentences about read holds.